File: informatics_msk/159_resistory.py
import sys


def solve() -> None:
    data = sys.stdin.read().strip().split()
    if not data:
        return
    n = int(data[0])
    C = float(data[1])
    caps = [float(x) for x in data[2 : 2 + n]]

    # dp[mask] — множество всех ёмкостей, которые можно получить,
    # используя в точности конденсаторы, соответствующие битам mask
    dp = [set() for _ in range(1 << n)]

    # базовый случай: один конденсатор
    for i in range(n):
        dp[1 << i].add(caps[i])

    # перебираем все непустые подмножества
    for mask in range(1, 1 << n):
        if mask & (mask - 1) == 0:  # только один конденсатор — уже посчитано
            continue
        lsb = mask & -mask  # младший бит (чтобы не учитывать симметричные разбиения)
        sub = (mask - 1) & mask
        while sub:
            if sub & lsb:  # подмножество содержит младший бит
                other = mask ^ sub  # оставшиеся конденсаторы
                s = dp[mask]
                for a in dp[sub]:
                    for b in dp[other]:
                        # параллельное соединение
                        s.add(a + b)
                        # последовательное соединение
                        s.add((a * b) / (a + b))
            sub = (sub - 1) & mask

    # ищем значение, ближайшее к C
    best_val = None
    best_diff = float("inf")
    for mask in range(1, 1 << n):
        for val in dp[mask]:
            diff = abs(val - C)
            if diff < best_diff:
                best_diff = diff
                best_val = val

    # выводим с шестью знаками после запятой
    print(f"{best_val:.6f}")


if __name__ == "__main__":
    solve()

File: informatics_msk/159_resistory.cpp
#include <climits>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <set>
#include <vector>
#include <unordered_map>
using namespace std;


int main() {
    int n;
    double c;
    cin >> n >> c;
    vector<double> a(n);
    for (int i = 0; i < n; i++) cin >> a[i];
    unordered_map<int, set<double>> dp;

    for (int i = 0; i < n; i++) {
        dp[1 << i].insert(a[i]);
    }

    for (int mask = 0; mask < (1 << n); mask++) {
        if (!(mask & (mask - 1))) continue;
        int lsb = mask & -mask;

        int sub = (mask - 1) & mask;
        while (sub) {
            if (sub & lsb) {
                int other = mask ^ sub;
                for (auto i : dp[sub]) {
                    for (auto j : dp[other]) {
                        dp[mask].insert(i + j);
                        dp[mask].insert((i * j) / (i + j));
                    }
                }
            }
            sub = (sub - 1) & mask;
        }
    }

    double best_val, best_diff = INT_MAX;
    for (auto pair : dp) {
        for (auto num : pair.second) {
            if (abs(c - num) < best_diff) {
                best_diff = abs(c - num);
                best_val = num;
            }
        }
    }

    cout << fixed << setprecision(6) << best_val;

    return 0;
}
